### PR TITLE
facto of split, algorithm definition, consistency between step and de…

### PR DESCRIPTION
**Pull request by MockingMagician**
* Implementing contract for Residue
* Adding consistency between step and decimal to prevent unexpected behaviors
* Defining algorithm of residue->split
* Refacto of splitWith and WithoutStep to split with the common algorithm

### CODE REVIEW

File: src/Residue.php [hunk 1]
```diff
@@ -12,22 +12,21 @@
...
-final class Residue
+final class Residue implements ResidueInterface
 {
-    /**
-     * @var float|int
-     */
-    private $value;
-    private int $divider;
+    private float $value;
+    private bool $isNegative = false;
```
Comment (romainnorberg) [inline src/Residue.php]:
```suggestion
    private bool $isNegative;
```

The assignment can be safely removed as the constructor overrides it.

File: src/Residue.php [hunk 2]
```diff
@@ -69,120 +143,54 @@ public function decimal(int $decimal): self
...
 
-    public function toArray(): array
+    public function toArray(string $mode = self::SPLIT_MODE_ALLOCATE): array
     {
-        return iterator_to_array($this->split(), false);
+        return iterator_to_array($this->split($mode), false);
     }
 
-    public function isNegative(float $value): bool
+    protected function isNegative(float $value): bool
```
Comment (romainnorberg) [inline src/Residue.php]: Since the class is final, the member can be declared private

File: src/Residue.php [hunk 3]
```diff
@@ -69,120 +143,54 @@ public function decimal(int $decimal): self
...
-                $yield = round($part + $this->remainder(), $this->decimal);
-
-                yield $this->isNegative ? -$yield : $yield;
-
-                break;
-            }
-
-            yield $this->isNegative ? -$part : $part;
+        $strFloat = (string) $float;
+        if (false === mb_strpos($strFloat, '.')) {
```
Comment (romainnorberg) [inline src/Residue.php:168]: Too bad there is no internal PHP function to do this job

File: src/Residue.php [hunk 4]
```diff
@@ -40,10 +39,83 @@ public static function create(float $value): self
...
+     *  TODO choice during implementation
+     *
+     *  - Define ns
+     *  - Define dNs
+     *
+     *  - Generate generator for parts
+     *
+     *  - Define rm according "to mode (so, split must be called)"
+     */
+    public function split(string $mode = self::SPLIT_MODE_ALLOCATE): Generator
```
Comment (romainnorberg) [inline src/Residue.php:76]: awesome! Much cleaner than actual 😮

File: src/Residue.php [hunk 5]
```diff
@@ -69,120 +143,54 @@ public function decimal(int $decimal): self
 
         $this->decimal = $decimal;
 
+        $this->step = 10 ** -$decimal;
```
Comment (romainnorberg) [inline src/Residue.php]: I've a strange behavior following the order of the method (step/decimal). 

here a test to illustrate:

```php

    /**
     * @test
     */
    public function it_should_split_test_case(): void
    {
        // work as expected
        $residue = Residue::create(7.315)
                    ->divideBy(3)
                    ->decimal(3)
                    ->step(0.050);

        $this->assertEquals([2.45, 2.45, 2.4], $residue->toArray());
        $this->assertEquals(0.015, $residue->getRemainder());

        // fail
        $residue = Residue::create(7.315)
                    ->divideBy(3)
                    ->step(0.050) // <-- step before decimal
                    ->decimal(3);

        $this->assertEquals([2.45, 2.45, 2.4], $residue->toArray());
        $this->assertEquals(0.015, $residue->getRemainder());
        /*
             * Failed asserting that two arrays are equal.
            --- Expected
            +++ Actual
            @@ @@
            Array (
            -    0 => 2.45
            -    1 => 2.45
            -    2 => 2.4
            +    0 => 2.439
            +    1 => 2.438
            +    2 => 2.438
            )
         */
    }
```

maybe don't overwrite step if it was previously defined

File: src/Residue.php [hunk 6]
```diff
@@ -40,10 +39,83 @@ public static function create(float $value): self
...
+     *  - Dv the divider
+     *  - Dc the number of decimals
+     *  - S the step size.
+     *
+     * This function is described by this equations :
+     * [
+     *  - S = S || 10^(-Dc)
+     *  - Dc = Dc || GetNumberOfDecimalOfStep(S)
+     *
+     *  - rm + p¹ + p² + ... p^n = N (where rm the reminder, p a part and n in range [1, Dv])
```
Comment (romainnorberg) [inline src/Residue.php]: reminder => remainder 

I hesitate between `remainder` and `leftover` _(adj. remaining/surplus)_ I don't know which is the best

File: tests/ResidueSplitTest.php [hunk 7]
```diff
@@ -14,286 +14,458 @@
...
+            2, // divide
+            null, // decimal value
+            0.03, // step
+            [
+                1.44,
+                1.44,
+            ], // expected result
+            0.03, // reminder
         ];
     }
```
Comment (romainnorberg) [inline tests/ResidueSplitTest.php:519]: do you think it's relevant to add a this test:

```php

    /**
     * @test
     */
    public function it_should_split_returns_iterator(): void
    {
        $residue = Residue::create(100)->divideBy(3);

        $this->assertSame([33.34, 33.33, 33.33], iterator_to_array($residue->split(), false));
    }
```

File: composer.json [hunk 8]
```diff
@@ -30,7 +30,14 @@
     "authors": [
         {
             "name": "Romain Norberg",
-            "email": "romainnorberg@gmail.com"
+            "email": "romainnorberg@gmail.com",
+            "role": "Maintainer, Owner, Developer"
+        },
+        {
+            "name": "Mocking Magician",
```
Comment (romainnorberg) [inline composer.json:37]: ✅  I had to done this before. sorry

**Review by romainnorberg — COMMENTED** [review]
it's really nice work, much cleaner and more robust than the current one

here is a first code review

thanks

File: src/Residue.php (hunk 6 above)
Comment (MockingMagician) [reply src/Residue.php]: > reminder => remainder

Yeah a bit of typo, well spoted. It is in the sense that I need to be reminded that there is a left over... :wink: 

Both are understandable but remainder is more accurate in mathematical context. I mind it was the good choice.

File: src/Residue.php (hunk 1 above)
Comment (MockingMagician) [reply src/Residue.php]: yes, cleary useless

File: src/Residue.php (hunk 2 above)
Comment (MockingMagician) [reply src/Residue.php]: Yeap !

File: src/Residue.php (hunk 3 above)
Comment (MockingMagician) [reply src/Residue.php:168]: Yes, I hope we'll have more mathematically oriented functions and better handling of floating point numbers in future versions of PHP...

File: src/Residue.php (hunk 4 above)
Comment (MockingMagician) [reply src/Residue.php:76]: Thanks !

File: tests/ResidueSplitTest.php (hunk 7 above)
Comment (MockingMagician) [reply tests/ResidueSplitTest.php:519]: I'm not sure to see the interest here because this is the behavior of toArray which is called for each test...

It seems to me that it's a bit like testing the behavior of the iterator_to_array function ...

The question is, can we write a test that would allow to fail the expected result with iterator_to_array ?

File: composer.json (hunk 8 above)
Comment (MockingMagician) [reply composer.json:37]: Don't worry, one way or another I participate of my own free will ;-)

File: src/Residue.php (hunk 5 above)
Comment (MockingMagician) [reply src/Residue.php]: The trick here is that step is set to 0.05 so it doesn't make sense that when decimal is set to 3 the behavior is like that. Since 0.001 is smaller than 0.05 anyway so can't add any more precision...

step should actually only be rewritten if decimal is strictly greather than step...

because if we do opposite :

```php
$residue = Residue::create(7.315)
                    ->divideBy(3)
                    ->step(0.050) // <-- step before decimal
                    ->decimal(1);

//One should expect a result of this type :

$this->assertEquals([2.5, 2.4, 2.4], $residue->toArray());
$this->assertEquals(0.015, $residue->getRemainder());
```

What do you think about it?

Comment (MockingMagician) [reply src/Residue.php]: > @romainnorberg maybe : don't overwrite step if it was previously defined
> @MockingMagician : step should actually only be rewritten if decimal is strictly greather than step...

I have pushed new implementation in that way that take care of both, you can know give it a try, if you find some edge cases to fix, it will be fine!

**Comment by MockingMagician** [comment]
When all the logic will be fine, I will add more set of tests for split... just to be sure. 
Too, I will clean up comments and the definition of the algorithm.

File: src/Residue.php (hunk 5 above)
Comment (romainnorberg) [reply src/Residue.php]: yeah @MockingMagician it's work well 👍 Behavior is ok for me

File: tests/ResidueSplitTest.php (hunk 7 above)
Comment (romainnorberg) [reply tests/ResidueSplitTest.php:519]: You're right, testing split is not relevant here _(split() output is typed)_.

Another proposition on the tests, would we not test the returned total? _(sum values + remainder)_:

```php
$this->assertEquals($value, array_sum($residue->toArray()) + $residue->getRemainder());
```

Comment (MockingMagician) [reply tests/ResidueSplitTest.php:519]: > Another proposition on the tests, would we not test the returned total? _(sum values + remainder)_:
> 
> ```
> $this->assertEquals($value, array_sum($residue->toArray()) + $residue->getRemainder());
> ```

Just great! It freezes the box perfectly, no chance of leakage while writing the tests.

**Comment by MockingMagician** [comment]
Clean up done, ready for last review... :rocket:

**Comment by romainnorberg** [comment]
Hi @MockingMagician everything seems ok 👌 I've updated the readme file and fix one typo. Ready for merge & release if you agree

File: src/Contracts/ResidueInterface.php [hunk 9]
```diff
@@ -31,15 +31,15 @@ public function decimal(int $decimal): self;
      *
      * SPLIT_MODE_ALLOCATE is default and try to allocate the maximum of the value according to step
      * example:
-     *   100/33 with step 1 will split into [34, 33, 33]
+     *   100/3 with step 1 will split into [34, 33, 33]
```
Comment (MockingMagician) [inline src/Contracts/ResidueInterface.php:34]: well spotted! :+1:

**Comment by MockingMagician** [comment]
Looks good, ready when you are :clap:

**Comment by romainnorberg** [comment]
Released https://github.com/romainnorberg/residue/releases/tag/v0.3.2 thanks @MockingMagician 

I will update https://residue-online-php-tester.herokuapp.com/ soon
